File: app/screenshot.py
#!/opt/venv/bin/python3

from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager

def take_screenshot(html_file, output_file):
    options = Options()
    options.headless = True
    options.add_argument("--window-size=1170,2532")  # iPhone 14 resolution
    options.add_argument("--no-sandbox")  # Bypass OS security model
    options.add_argument("--disable-dev-shm-usage")  # Overcome limited resource problems

    driver = webdriver.Chrome(options=options)

    driver.get(f"file://{html_file}")

    # Take a screenshot of the entire page
    element = driver.find_element_by_tag_name('body')
    element.screenshot(output_file)

    # Close the browser
    driver.quit()

if __name__ == "__main__":
    html_file = 'deals.html'  # Path to your HTML file
    output_file = 'screenshot.png'  # Output file path
    take_screenshot(html_file, output_file)
    print(f"Screenshot saved to {output_file}")

File: app/screenshot.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');
const url = require('url');

async function takeScreenshot(filePath, outputPath) {
    const browser = await puppeteer.launch({
        headless: "new",
        executablePath: '/usr/bin/chromium',
        args: ['--no-sandbox', '--disable-setuid-sandbox']
    });

    const page = await browser.newPage();
    await page.setViewport({ width: 1256, height: 800 });

    const fileUrl = url.pathToFileURL(path.resolve(filePath)).toString();
    await page.goto(fileUrl, { waitUntil: 'networkidle2' });
    await page.screenshot({ path: outputPath, fullPage: true });
    await browser.close();
}

function getLatestOutputFolder(baseFolder) {
    const dirs = fs.readdirSync(baseFolder, { withFileTypes: true })
        .filter(dirent => dirent.isDirectory() && dirent.name.startsWith('output_'))
        .map(dirent => ({ name: dirent.name, time: fs.statSync(path.join(baseFolder, dirent.name)).mtime }))
        .sort((a, b) => b.time - a.time);
    return dirs.length > 0 ? path.join(baseFolder, dirs[0].name) : null;
}

async function screenshotLatestOutput(baseFolder) {
    const latestFolder = getLatestOutputFolder(baseFolder);
    if (!latestFolder) {
        console.log('No output folder found.');
        return;
    }

    const htmlFiles = fs.readdirSync(latestFolder)
        .filter(file => file.endsWith('.html'));

    for (const file of htmlFiles) {
        const filePath = path.join(latestFolder, file);
        const outputPath = path.join(latestFolder, `${path.parse(file).name}.png`);

        // Skip if screenshot already exists
        if (fs.existsSync(outputPath)) {
            console.log(`Screenshot already exists for ${file}, skipping...`);
            continue;
        }

        console.log(`Taking screenshot of ${file}`);
        await takeScreenshot(filePath, outputPath);
    }
}

const baseFolder = 'publish'; // Adjust to your base folder path
screenshotLatestOutput(baseFolder)
    .then(() => console.log('All screenshots taken'))
    .catch(err => console.error('Error:', err));
